Android-Sample/app/src/main/java/com/opentok/qualitystats/sample/MainActivity.java

```diff
@@ -2,407 +2,205 @@
+                                + "Audio Packet Lost Ratio  " + stats.getAudioPacketLostRatio() * 100 + "% \n"
+                                + "Video Packet Lost Ratio  " + stats.getVideoPacketLostRatio() * 100 + "%\n"
+                                + "Current Round Trip Time: " + stats.getCurrentRoundTripTimeMs() + " ms";
+
+                        statsTextView.setText(sentVideoInfo);
+
+                        Log.d(LOGTAG, "---------------------------------------------------------------");
+                        Log.d(LOGTAG, "Sent Video Bitrate: " + stats.getSentVideoBitrateKbps() + " Kbps");
+                        Log.d(LOGTAG, "Sent Audio Bitrate: " + stats.getSentAudioBitrateKbps() + " Kbps");
+                        Log.d(LOGTAG, "Received Audio Bitrate: " + stats.getReceivedAudioBitrateKbps() + " Kbps");
+                        Log.d(LOGTAG, "Received Video Bitrate: " + stats.getReceivedVideoBitrateKbps() + " Kbps");
+                        Log.d(LOGTAG, "Current Round Trip Time: " + stats.getCurrentRoundTripTimeMs() + " ms");
+                        Log.d(LOGTAG, "Available Outgoing Bitrate: " + stats.getAvailableOutgoingBitrate() + " bps");
+                        Log.d(LOGTAG, "Audio Packet Lost Ratio  " + stats.getAudioPacketLostRatio() * 100 + "%");
+                        Log.d(LOGTAG, "Video Packet Lost Ratio  " + stats.getVideoPacketLostRatio() * 100 + "%");
+                        Log.d(LOGTAG, "Jitter: " + stats.getJitter());
+                        Log.d(LOGTAG, "Quality Limitation Reason: " + stats.getQualityLimitationReason());
+                        Log.d(LOGTAG, "Sent video resolution: " + stats.getSentVideoResolution());
+                        Log.d(LOGTAG, "Received video resolution: " + stats.getReceivedVideoResolution());
+                        Log.d(LOGTAG, "Scalable video ? " + stats.isScalableVideo());
+                        Log.d(LOGTAG, "---------------------------------------------------------------");
+
+                        availableOutgoingNitrateResult.add(stats.getAvailableOutgoingBitrate());
+                        sentVideoBitrateResults.add(stats.getSentVideoBitrateKbps());
+                        sentAudioBitrateResults.add(stats.getSentAudioBitrateKbps());
+
+                        updateChart(videoUploadSpeedChart, sentVideoBitrateResults, "Sent Video Bitrate");
+                        updateChart(audioUploadSpeedChart, sentAudioBitrateResults, "Sent Audio Bitrate");
+                        updateChart(availableOutgoingBitrateChart,
+                                availableOutgoingNitrateResult, "AvailableOutgoingBitrate");
+                    }
+
+                    @Override
+                    public void onError(String error) {
+                        Log.d(LOGTAG, "Error " + error);
+                        showErrorPopup(error);
+                    }
+                });
+        // Start the quality test
+        networkQualityTest.startTest();
```

Comment:
Should we only start the test when we start publishing? What if the user does not provide the permissions, or takes a lot of time to do so, and the test finishes without even having published? (Based on what I understood, if not explicitly stopped the test only takes 30s)

Reply:
I moved permission to the MainActivity and only run the test now if the user accepted the permissions.